HIRS_ProvisionerTPM2/src/Process.cpp

```diff
@@ -131,3 +142,68 @@ string Process::run(const string& executable,
     }
     return str;
 }
+
+/**
+ * Static function to check if a specified process is currently running in the
+ * local environment.
+ *
+ * @param executable the executable to check is running
+ * @param checkCmdline a flag to check the complete cmdline commands
+ *  rather than just search for the process name, default false
+ * @return true, if executable is found to be running / false, otherwise
+ */
+bool Process::isRunning(const string& executable, bool checkCmdline) {
+    // If the executable is the empty string, return false
```

Comment:
You could replace the contents of this method with the following line and get the functionality you want. You can still have your static method so the code that uses it is easy to read. You just get less lines of code and the ability to rely on the fact that `pgrep` has had years of testing. Less code == less maintenance $$
```suggestion
    return Process("pgrep", "-c " + executable).run() == 0;
```

Reply:
In the case of this project, that last sentiment has burned us before, and my hesitance to bring in additional tooling to handle an otherwise simple task is reflective of that. That being said, we each clearly have our own amount of skepticism of either approach that is at a cross-purpose with one another. We'll go with your route on the basis of moving things along.

Reply:
I just experimented with the proposed `pgrep` solution and it doesn't work. Since we're invoking the tool `pgrep` via C++ and C++ (and likely C) uses the `sh` command to invoke commands via `popen`, then it will trivially always have a process running for the `cmdline` usage.

I'm putting the old version back in and this is being marked as resolved.

Reply:
![image](https://user-images.githubusercontent.com/36857962/48792634-bc31f000-ecc2-11e8-9783-91a0f1d8f822.png)

Here's the failing test case demonstrating the aforementioned issue.

Reply:
The new test above fails when testing a capability involving searching `/proc/PID/cmdline`. Neither `pgrep` nor your implementation of it correctly handles this file they way we want. The `/proc/PID/cmdline` file contains more than the executable name. It includes the fully qualified executable path and all of the arguments -- either of which could contain the character string you are receiving as `executable`. Furthermore, `isRunning()` currently passes the responsibility of deciding which value to use for `checkCmdline` to the user of the API along with the risks associated with passing `true` to the current implementation. There are two good fixes here:

1.  Since the two uses of `isRunning()` you need to complete this issue are both below the 15 character limit, they don't require you to write an implementation that correctly handles searching `/proc/PID/cmdline`. That means you can easily eliminate `checkCmdline` and its current logic, disallow `executable` from containing paths, and compare `executable` to `/proc/PID/stat`. `pgrep` can do the later for you using the single line I provided above -- even with C++ calling `pgrep` via `sh -c`. Either way, given that the 15 character limit is a function of the `stat` file and not of `pgrep` or your pgrep implementation, you should log a warning that `isRunning()` could be getting a false negative when `executable` is more than 15 characters and is being compared to `proc/PID/stat`.

2. If you really want to provide the additional capability to handle the case where `executable` is over 15 characters, I still recommend removing the `checkCmdline` parameter. If the 15 character limit comes into play, `isRunning()` should be able to determine if looking at `/proc/PID/cmdline` is necessary.  If it is necessary, you could modify your implementation of `pgrep` to take the first token of `/proc/PID/cmdline` after tokenizing on whitespace, extract the filename at the end of that token, and compare it to `executable`. This would provide an improvement over `pgrep` rather than just reimplementing it. If you are going to allow `executable` to contain absolute or relative paths, a full solution must deal with a lot more (e.g., differing but equivalent paths).

Reply:
Alright, so taking the first approach, it's still not quite as simple as suggested and there are still edge cases to consider regardless of this approach.

Specifically, consider the following.

![image](https://user-images.githubusercontent.com/36857962/49034276-745a0f80-f17f-11e8-92a8-7a0cc4693f0e.png)

If you are comfortable moving forward with this implementation, then I'll upload it, but otherwise, I'd rather not overwrite everything, and then have to switch back later.

My point here is that even this seemingly simplified approach of just running it as a Process has Edge Cases. It's whether you'd rather risk those edge cases as a race condition or as a potential Security liability.

Reply:
Alright after a quick outside discussion, I've pushed the above code as pictured. I also pushed the commit and we can see about leveraging Github's automatic rebasing capability to handle squashing all the commits into one.